drivers/gpio/gpio_sam0.c

```diff
@@ -172,52 +117,147 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
 			  u32_t *value)
 {
 	const struct gpio_sam0_config *config = DEV_CFG(dev);
-	u32_t bits;
 
-	if (access_op != GPIO_ACCESS_BY_PIN) {
-		/* TODO(mlhx): support GPIO_ACCESS_BY_PORT */
-		return -ENOTSUP;
-	}
+	*value = (config->regs->IN.reg & BIT(pin)) != 0;
+	return 0;
+}
 
-	bits = config->regs->IN.bit.IN;
-	*value = (bits >> pin) & 1;
+static int gpio_sam0_port_get_raw(struct device *dev,
+				  gpio_port_value_t *value)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
 
+	*value = config->regs->IN.reg;
 	return 0;
 }
 
-#ifdef CONFIG_SAM0_EIC
+static int gpio_sam0_port_set_masked_raw(struct device *dev,
+					 gpio_port_pins_t mask,
+					 gpio_port_value_t value)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+	u32_t out = config->regs->OUT.reg;
 
-int gpio_sam0_manage_callback(struct device *dev,
-			      struct gpio_callback *callback, bool set)
+	config->regs->OUT.reg = (out & ~mask) | (value & mask);
+
+	return 0;
+}
+
+static int gpio_sam0_port_set_bits_raw(struct device *dev,
+				       gpio_port_pins_t pins)
 {
-	struct gpio_sam0_data *const data = DEV_DATA(dev);
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
 
-	return gpio_manage_callback(&data->callbacks, callback, set);
+	config->regs->OUTSET.reg = pins;
+	return 0;
 }
 
-int gpio_sam0_enable_callback(struct device *dev, int access_op, u32_t pin)
+static int gpio_sam0_port_clear_bits_raw(struct device *dev,
+					 gpio_port_pins_t pins)
 {
 	const struct gpio_sam0_config *config = DEV_CFG(dev);
 
-	if (access_op != GPIO_ACCESS_BY_PIN) {
-		return -ENOTSUP;
-	}
+	config->regs->OUTCLR.reg = pins;
+	return 0;
+}
 
-	return sam0_eic_enable_interrupt(config->id, pin);
+static int gpio_sam0_port_toggle_bits(struct device *dev,
+				      gpio_port_pins_t pins)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+
+	config->regs->OUTTGL.reg = pins;
+	return 0;
 }
 
-int gpio_sam0_disable_callback(struct device *dev, int access_op, u32_t pin)
+#ifdef CONFIG_SAM0_EIC
+
+static int gpio_sam0_pin_interrupt_configure(struct device *dev,
+					     unsigned int pin,
+					     enum gpio_int_mode mode,
+					     enum gpio_int_trig trig)
 {
 	const struct gpio_sam0_config *config = DEV_CFG(dev);
+	PortGroup *regs = config->regs;
+	PORT_PINCFG_Type pincfg = {
+		.reg = regs->PINCFG[pin].reg,
+	};
+	enum sam0_eic_trigger trigger;
+	int rc = 0;
+
```

Comment:
I think we only need to check `INEN`; interrupts should still work in bidirectional mode.

Reply:
The documentation could be more clear about it, however we do need to configure pin as function A in pinmux. This disconnects the PORT module. So I doubt the bidirectional mode would work. Also your earlier comments confirms that

> Although I've confirmed that setting and clearing the pin results in a change in the input value before gpio_pin_interrupt_configure(), while the interrupts are configured no changes to the input value can be detected through the GPIO interface

Reply:
Correct, you can't do bidirectional mode when the interrupts are active, but you can have the GPIO configured that way underneath so that when interrupts are turned off the pin returns to bidirectional mode.  So `INEN` is enough to gate entry to interrupt mode.  I've confirmed this by running the tests with the input pin configured bidirectional.

Reply:
> you can have the GPIO configured that way underneath so that when interrupts are turned off the pin returns to bidirectional mode.

We shouldn't do it. The input shouldn't suddenly become an output unless explicitly configured via `gpio_pin_configure` function. That's dangerous.

Reply:
I don't understand.   The way this API model works is (a) configure the GPIO (direction, pull, debounce), then (b) configure (and enable) interrupts.

Deconfiguring (disabling) interrupts should restore the underlying GPIO configuration, whether it was input-only, output only (where that works), or bidirectional.  Do you think it should do something else?

Reply:
the (b) cannot change output mode to input and vice versa. The `gpio_pin_interrupt_configure` function should only configure interrupts.

Reply:
Um, yeah, it is.  DIRSET and INEN, works fine.

Reply:
Does the driver pass 1 pin testcase?

Reply:
> Does the driver pass 1 pin testcase?

Everything but the interrupt tests.

The 2-pin test passes when the input GPIO is configured as disconnected.  So on this device the GPIO is completely disabled (except possibly for the effect of pull settings) when interrupts are enabled.  The check I added for `INEN` being set is wrong: you don't need the input buffer enabled support external interrupts.

So I'm not sure what to do.  The 1-pin test for interrupts simply can't work on this device, but that can't be detected without improperly rejecting GPIO/interrupt configurations that are supported.  Perhaps the 1-pin test should skip the interrupt section if `IS_ENABLED(CONFIG_GPIO_SAM0)`.

Reply:
> The 2-pin test passes when the input GPIO is configured as disconnected.

OK, that was not obvious. According to the documentation `INEN` has to be enabled for the input signal to be routed to other modules. But documentation is not always perfect.

> So I'm not sure what to do. The 1-pin test for interrupts simply can't work on this device, but that can't be detected without improperly rejecting GPIO/interrupt configurations that are supported. 

If the 1-pin testcase doesn't work then clearly some GPIO/interrupt configurations are not supported. As I stated already `gpio_pin_interrupt_configure` cannot change pin mode from input to output or vice versa. This is reserved for `gpio_pin_configure` function. If GPIO pin is configured as an output and then `gpio_pin_interrupt_configure` connects it to EIC module by means of pinmux function A the pin becomes an input.

Reply:
> If the 1-pin testcase doesn't work then clearly some GPIO/interrupt configurations are not supported. As I stated already `gpio_pin_interrupt_configure` cannot change pin mode from input to output or vice versa. This is reserved for `gpio_pin_configure` function. If GPIO pin is configured as an output and then `gpio_pin_interrupt_configure` connects it to EIC module by means of pinmux function A the pin becomes an input.

I don't think that model is correct for this driver. `gpio_pin_interrupt_configure()` will transform a disconnected GPIO to an "input" even if `INEN` is not set.

However, I'm done arguing for consistency.  I've pushed a commit that performs the validation you want, and now the 1-pin interrupt test is skipped on this platform.

Reply:
> `gpio_pin_interrupt_configure()` will transform a disconnected GPIO to an "input" even if `INEN` is not set.

This isn't allowed. The `gpio_pin_interrupt_configure()` can't change pin drive settings. It should only verify that the pin drive settings are compatible with the interrupt mode, i.e.  that the pin is an input. This is the case now, finally, so that's good.

We need to change however the comment you've added to the code as it is not correct and confusing. I propose to write something like:
"Enabling pin interrupt requires disconnecting it from the I/O pin controller (PORT) module and connecting to the External Interrupt Controller (EIC). As a result interrupts are only supported if the pin is configured in input mode."

Reply:
Is the new wording acceptable?

Reply:
Looks good!